File: ts/test/entity/v1_lookup/V1LookupEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { FreeMusicSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('V1LookupEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when FREE_MUSIC_TEST_LIVE=TRUE.
  afterEach(liveDelay('FREE_MUSIC_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = FreeMusicSDK.test()
    const ent = testsdk.V1Lookup()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.FREE_MUSIC_TEST_LIVE
    for (const op of ['list', 'load']) {
      if (maybeSkipControl(t, 'entityOp', 'v1_lookup.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set FREE_MUSIC_TEST_V1_LOOKUP_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let v1_lookup_ref01_data = Object.values(setup.data.existing.v1_lookup)[0] as any

    // LIST
    const v1_lookup_ref01_ent = client.V1Lookup()
    const v1_lookup_ref01_match: any = {}
    v1_lookup_ref01_match['api_key'] = setup.idmap['api_key01']

    const v1_lookup_ref01_list = (await v1_lookup_ref01_ent.list(v1_lookup_ref01_match)).map((e: any) => e.data())



  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/v1_lookup/V1LookupTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = FreeMusicSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['v1_lookup01','v1_lookup02','v1_lookup03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['FREE_MUSIC_TEST_V1_LOOKUP_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'FREE_MUSIC_TEST_V1_LOOKUP_ENTID': idmap,
    'FREE_MUSIC_TEST_LIVE': 'FALSE',
    'FREE_MUSIC_TEST_EXPLAIN': 'FALSE',
    'FREE_MUSIC_APIKEY': '',
  })

  idmap = env['FREE_MUSIC_TEST_V1_LOOKUP_ENTID']

  const live = 'TRUE' === env.FREE_MUSIC_TEST_LIVE

  if (live) {
    client = new FreeMusicSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
        apikey: env.FREE_MUSIC_APIKEY,
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {}
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.FREE_MUSIC_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  

File: go/test/v1_lookup_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/free-music-sdk/go"
	"github.com/voxgig-sdk/free-music-sdk/go/core"

	vs "github.com/voxgig-sdk/free-music-sdk/go/utility/struct"
)

func TestV1LookupEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.V1Lookup(nil)
		if ent == nil {
			t.Fatal("expected non-nil V1LookupEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := v1_lookupBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "v1_lookup." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FREEMUSIC_TEST_V__LOOKUP_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		v1LookupRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.v1_lookup", setup.data)))
		var v1LookupRef01Data map[string]any
		if len(v1LookupRef01DataRaw) > 0 {
			v1LookupRef01Data = core.ToMapAny(v1LookupRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = v1LookupRef01Data

		// LIST
		v1LookupRef01Ent := client.V1Lookup(nil)
		v1LookupRef01Match := map[string]any{
			"api_key": setup.idmap["api_key01"],
		}

		v1LookupRef01ListResult, err := v1LookupRef01Ent.List(v1LookupRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, v1LookupRef01ListOk := v1LookupRef01ListResult.([]any)
		if !v1LookupRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", v1LookupRef01ListResult)
		}

		// LOAD
		v1LookupRef01MatchDt0 := map[string]any{}
		v1LookupRef01DataDt0Loaded, err := v1LookupRef01Ent.Load(v1LookupRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if v1LookupRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func v1_lookupBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "v1_lookup", "V1LookupTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read v1_lookup test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse v1_lookup test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"v1_lookup01", "v1_lookup02", "v1_lookup03", "api_key01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FREEMUSIC_TEST_V__LOOKUP_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FREEMUSIC_TEST_V__LOOKUP_ENTID": idmap,
		"FREEMUSIC_TEST_LIVE":      "FALSE",
		"FREEMUSIC_TEST_EXPLAIN":   "FALSE",
		"FREEMUSIC_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["FREEMUSIC_TEST_V__LOOKUP_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FREEMUSIC_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["FREEMUSIC_APIKEY"],
			},
			extra,
		})
		client = sdk.NewFreeMusicSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FREEMUSIC_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FREEMUSIC_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
